neps/nep-0611.md

```diff
@@ -0,0 +1,233 @@
+transactions are valid. At a high level, they are:
+* **Access Key vs Gas Key**: In addition to Access Keys, we introduce a second kind of signing key called the Gas Key. A gas key can be funded with NEAR, and when issuing a transaction using a gas key, the transaction only consumes gas from the gas key, not from the account.
+* **Pending Transaction Queue**: Chunk producers keep track of pending (accepted into a block, but not 
+  executed) transactions, and ensures that transactions sent with access keys are limited in parallelism,
+  whereas transactions sent with gas keys are limited to the available gas in the gas key.
+
+We will now specify how exactly they work.
+
+### Definition of "Pending Transactions"
+In a model where execution follows but lags behind consensus, there are transactions which are accepted
+into consensus and thus committed to be executed in the near future, but are not yet executed. This set
+of transactions is called the *pending transactions*. We always discuss this in the context of one shard.
+
+Note that there are two slightly different ways to treat this definition, depending on how exactly the
+"execution head" (how far the execution has caught up) is defined:
+* It can be defined locally as the progress of execution at a node, but this will be different between
+  nodes.
+* It can be defined deterministically as the last block whose execution result is certified by consensus.
+
+For the purpose of this NEP, we use the latter definition, so that the notion of pending transactions is
+consistent across all nodes and the determinination of what transactions are eligible to be included by
+a chunk producer can be verified -- even though we do not plan to implement this verification right now.
+
+Another note is that the notion of pending transactions is anchored at a specific chunk that is being produced. In case of forks, we use the block that the chunk is being produced on top of to compute the
+set of pending transactions.
+
+Finally, this NEP does *not* depend on the implementation of SPICE. In the context where SPICE is not
+yet in effect, we consider the pending transactions queue to always be empty (despite technically being
+one chunk worth of transactions due to execution lagging one block behind in the current implementation),
+because we can always verify the validity of all transactions at the moment of inclusion.
+
+### Access Key vs Gas Key
+We introduce a new transaction version, `TransactionV2`, which is able to either specify an access key
+or a gas key. Any older transaction version is equivalent to a `TransactionV2` that specifies an access
+key.
+
+#### Access Key Parallelism Restriction
+We now restrict the ability to send multiple parallel pending transactions with Access Keys.
+
+Specifically, for any given account $A$ with any number of access keys, the total number of access key transactions in the pending transaction queue whose sender is $A$ cannot exceed $P_{\mathrm {max}}$, a
```

Comment:
Is this enough to prevent the attack described above? Instead of having one account they send many transactions to, it could be many accounts they send only a few transactions to each.

Reply:
The idea is that the amount of block space the attacker consumes vs amount of gas the attacker has to pay, i.e. the "waste amplification factor", is unbounded in the attack mentioned in the NEP, but only O(1) with a maximum parallelism. It's then still possible to attack, but the attack would cost a proportional amount of gas - just cheaper than it is today - which isn't a very great incentive to mount the attack (given that the attack is only a DoS).